src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php

```diff
@@ -112,20 +114,22 @@ public function getConfigTreeBuilder(): TreeBuilder
                     ->prototype('scalar')->end()
                 ->end()
                 ->arrayNode('trusted_hosts')
-                    ->beforeNormalization()->ifString()->then(static fn ($v) => $v ? [$v] : [])->end()
+                    ->castToArray(['string'])
+                    ->beforeNormalization()->ifArray()->then(static fn ($v) => [''] === $v ? [] : $v)->end()
                     ->prototype('scalar')->end()
                     ->defaultValue(['%env(default::SYMFONY_TRUSTED_HOSTS)%'])
                 ->end()
                 ->variableNode('trusted_proxies')
                     ->beforeNormalization()
-                        ->ifTrue(fn ($v) => 'private_ranges' === $v || 'PRIVATE_SUBNETS' === $v)
-                        ->then(fn () => IpUtils::PRIVATE_SUBNETS)
+                        ->ifString()
+                        ->then(static fn ($v) => 'private_ranges' === $v || 'PRIVATE_SUBNETS' === $v ? IpUtils::PRIVATE_SUBNETS : $v)
```

Comment:
Having to add extra conditions in the `then` part makes the code less readable IMO, by making it harder to figure out when things are actually changed.

If we need to keep an explicit string check to tell the code generator that the normalization supports only string input, we should probably add an `if*` method allowing to provide a closure taking a string as input, where the final `if` condition would be `is_string($v) && $$callback($v)` (and still recording `string` as the allowed type)